client/login/controller.js

```diff
@@ -159,3 +157,37 @@ export function redirectJetpack( context, next ) {
 	}
 	next();
 }
+
+export function setShouldSSRLogin( context, next ) {
+	if ( ! context.SSRValidators ) {
+		next(); // practically somehow the initSSRValidators was not initialized, so adding more is not necessary
+		return;
+	}
+
+	context.SSRValidators = Object.assign( context.SSRValidators, {
+		// Overrides the query validator, which by default would ask no parameters
+		query: () => {
+			const whitelistedQueryKeys = [ 'client_id', 'signup_flow', 'redirect_to' ];
+			const queryKeys = Object.keys( context.query );
+
+			// if there are any parameters, they must be ONLY the ones in the whitelist
+			return queryKeys.length === intersection( queryKeys, whitelistedQueryKeys ).length;
+		},
+
+		// When it exists, the redirect_to query parameter must satisfy a specific format
+		query__redirect_to: () => {
+			if ( ! context.query.redirect_to ) {
+				return true; // nothing to check
+			}
+
+			const redirectToDecoded = decodeURIComponent( context.query.redirect_to );
+			return (
+				redirectToDecoded.length < 120 &&
+				( redirectToDecoded.startsWith( 'https://wordpress.com/theme' ) ||
```

Comment:
It might be worth making a mental note of the URLs listed here. I think vertical LPs are coming soon that will include `https://wordpress.com/websites/` and that might be another we should add to the whitelist.